### PR TITLE
Response Name + Typed Tool Params

### PR DESCRIPTION
**Pull request by ethanleifer**
Lmk what you think @roodboi.

Would do this for the rest of param parsers so we don't have typos like in L21 of build function one

### CODE REVIEW

File: src/instructor.ts [hunk 1]
```diff
@@ -246,16 +251,20 @@ class Instructor {
...
+    }).definitions
 
-    this.log("JSON Schema from zod: ", jsonSchema)
+    this.log("JSON Schema from zod: ", modelJsonSchema)
 
     const definition = {
-      name: "response_model",
-      ...jsonSchema.definitions?.response_model
+      name: model_name,
+      description: response_model.description,
```
Comment (roodboi) [inline src/instructor.ts:266]: AH, I see now - we were passing the top level schema vs just the definitions - makes sense now - super curious though now as to why they still work - 

 I'd say the only thing would be we should prob jsut make sure we have definitions for sure, or maybe just throw explicitly if not here

something like:

```
    const { schema, name = "response_model", description } = response_model

    const { definitions } = zodToJsonSchema(schema, {
      name,
      errorMessages: true
    })

    if (!definitions || !definitions?.[name]) {
      console.warn("Could not extract json schema definitions from your schema", schema)
      throw new Error("Could not extract json schema definitions from your schema")
    }

    this.log("JSON Schema from zod: ", definitions)

    const definition = {
      name,
      description,
      ...definitions[name]
    }

    const paramsForMode = MODE_TO_PARAMS[this.mode](definition, params, this.mode)

    return {
      stream: false,
      ...paramsForMode
    }
    ```

Comment (roodboi) [reply src/instructor.ts:266]: oh or I guess no, I read that wrong - I guess I'm just missing where the typo was

**Comment by ethanleifer** [comment]
Yea agree with you on definitions

Is there a reason we still support functions api even though it's deprecated @jxnl

**Comment by roodboi** [comment]
TBH I think we should nuke it

**Comment by roodboi** [comment]
But I guess it depends on the models you are using right? Also not sure about anyscale?

**Comment by jxnl** [comment]
i think some open source models are behind, and might support functions before tools. i think a warning is good since theres actualy performance differences.

**Pull request closed without merging**